qt: Enable tabbing through labels

hebasto opened the pull request:
This PR:
- fixes a visual bug described in #14577
- improves UX by enabling moving focus through fields using keyboard (**Tab** and **Shift+Tab** as usual) on Overview tab of the main window and on Info tab of the Debug window (other tabs leaved untouched as tab order for them seems a bit entangled). Inspired by **laanwj**'s https://github.com/bitcoin/bitcoin/pull/14577#issuecomment-438652059

Also applied a style-fix from [#15220](https://github.com/bitcoin/bitcoin/pull/15220#issuecomment-456377390):
> please just do this the next time you're editing these files.

laanwj commented:
Concept ACK, thanks!

On Mon, Nov 26, 2018, 17:21 Hennadii Stepanov <notifications@github.com
wrote:

> Inspired by @laanwj <https://github.com/laanwj>'s #14577 (comment)
> <https://github.com/bitcoin/bitcoin/pull/14577#issuecomment-438652059>
>
> This PR improves UX by enabling moving focus through QLabel fields using
> keyboard (*Tab* and *Shift+Tab* as usual).
>
> On the way this PR fixes a visual bug described in #14577
> <https://github.com/bitcoin/bitcoin/pull/14577> and improves UX for help
> windows.
>
> cc: @promag <https://github.com/promag> @fanquake
> <https://github.com/fanquake> @jonasschnelli
> <https://github.com/jonasschnelli>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/bitcoin/bitcoin/pull/14810
> Commit Summary
>
>    - Add event filter for focused QLabel objects
>    - Enable tabbing through labels for Debug Window
>    - Enable tabbing through labels for Overview view
>    - Improve UX for help dialogs
>    - Enable tabbing through labels for Send view
>    - Enable tabbing through labels for Coin Selection
>
> File Changes
>
>    - *M* src/qt/bitcoin.cpp
>    <https://github.com/bitcoin/bitcoin/pull/14810/files#diff-0> (2)
>    - *M* src/qt/forms/coincontroldialog.ui
>    <https://github.com/bitcoin/bitcoin/pull/14810/files#diff-1> (55)
>    - *M* src/qt/forms/debugwindow.ui
>    <https://github.com/bitcoin/bitcoin/pull/14810/files#diff-2> (192)
>    - *M* src/qt/forms/helpmessagedialog.ui
>    <https://github.com/bitcoin/bitcoin/pull/14810/files#diff-3> (11)
>    - *M* src/qt/forms/overviewpage.ui
>    <https://github.com/bitcoin/bitcoin/pull/14810/files#diff-4> (60)
>    - *M* src/qt/forms/sendcoinsdialog.ui
>    <https://github.com/bitcoin/bitcoin/pull/14810/files#diff-5> (48)
>    - *M* src/qt/guiutil.cpp
>    <https://github.com/bitcoin/bitcoin/pull/14810/files#diff-6> (21)
>    - *M* src/qt/guiutil.h
>    <https://github.com/bitcoin/bitcoin/pull/14810/files#diff-7> (19)
>    - *M* src/qt/overviewpage.cpp
>    <https://github.com/bitcoin/bitcoin/pull/14810/files#diff-8> (6)
>    - *M* src/qt/overviewpage.h
>    <https://github.com/bitcoin/bitcoin/pull/14810/files#diff-9> (1)
>    - *M* src/qt/rpcconsole.cpp
>    <https://github.com/bitcoin/bitcoin/pull/14810/files#diff-10> (3)
>
> Patch Links:
>
>    - https://github.com/bitcoin/bitcoin/pull/14810.patch
>    - https://github.com/bitcoin/bitcoin/pull/14810.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/14810>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHutqZNtoB4ikOhZ9AnLhCNU6b3U0-Yks5uzBUegaJpZM4YzXb1>
> .
>

jonasschnelli commented:
The only downside with this is, that you can't select parts of the text of a label.

laanwj commented:
@jonasschnelli Yeah. Possibly we're getting too deep in shed-painting territory here. I'll leave the decision what to do here up to you.

hebasto commented:
@jonasschnelli @laanwj 
> The only downside with this is, that you can't select parts of the text of a label.

That has been made intentionally.
Rationale: some labels are modified by `setText()` and any user selected text will vanish.
In such circumstances, only two options are consistent: all-selected or none-selected.
To make a focused label clearly visible to a user the former option has been chosen.

hebasto commented:
Rebased.

hebasto commented:
@promag @Sjors would you mind reviewing this PR?

promag commented:
@hebasto sure, on my list now.

Sjors commented:
Concept ACK.

Unfortunately this is a bit difficult to fully test on macOS because the menu keyboard shortcuts don't work (#13832), so you need the mouse for most navigation anyway.

I'm confused why `forms/overviewpage.ui` is modified; what's the point of being able to focus on the amounts? Is it so they're easier to copy-paste? Or is it an accessibility feature?

I was already able to navigate through the peer list. Now with this change I can select and copy individual items.

When I tab through the Send screen, it only goes through Pay to, label, amount, [nothing] and then the balance field.

On the receive screen it skips the BTC  dropdown field, the payment request button and the erase button. I can go through the list of historical payment request, but I can't access their view and delete buttons.

On the transactions page I can't access the the three filter dropdowns (on the left), and also not the Export button.

I also can't reach the wallet switcher.

hebasto commented:
@Sjors Thank you for your review and valuable input.

> I'm confused why `forms/overviewpage.ui` is modified; what's the point of being able to focus on the amounts? Is it so they're easier to copy-paste? Or is it an accessibility feature?

Yes, it is an accessibility feature mostly. Ref: https://github.com/bitcoin/bitcoin/pull/14577#issuecomment-435073188.

All others points will be addressed soon.

Sjors commented:
In that case, is there a way to trigger the tooltip by keyboard too?

promag commented:
Just wondering if changing to `QLineEdit` with `readOnly=true` makes sense? At least visually it's clear?

promag commented:
BTW I've noticed that #12616 no longer works.

practicalswift commented on src/qt/guiutil.h:
```diff
@@ -143,6 +143,25 @@ namespace GUIUtil
+        Q_OBJECT
+
+    public:
+        explicit FocusedLabelFilter(QObject* parent = 0);
```
Use `nullptr` instead of `0` :-)

jonasschnelli commented:
Tested this again and the lack of ability to select parts of an amount label (say the amount without the unit `BTC`) makes me NACK this as it is now.

hebasto committed:
qt: Add HighlightLabelWidget

hebasto committed:
qt: Make Information tab accessible via keyboard

hebasto committed:
qt: Make Overview tab accessible via keyboard

hebasto commented:
@promag 
> Just wondering if changing to `QLineEdit` with `readOnly=true` makes sense? At least visually it's clear?

This approach does not work for me (background color issues), see #17898.

> BTW I've noticed that #12616 no longer works.

Fixed in the latest push.

---

@jonasschnelli 

> The only downside with this is, that you can't select parts of the text of a label.

> Tested this again and the lack of ability to select parts of an amount label (say the amount without the unit `BTC`) makes me NACK this as it is now.

Fixed in the latest push.

promag commented:
@hebasto thanks for trying my suggestion in #17898. Will try and review both.

hebasto commented:
The OP has been updated.

hebasto commented:
@Sjors 
> In that case, is there a way to trigger the tooltip by keyboard too?

It seems possible. This could be addresses in a follow up PR ;)

hebasto commented:
Invariants are forced by `HighlightLabelWidget` ctor.

hebasto commented:
@jonasschnelli Mind re-reviewing?

hebasto commented:
Rebased d356340b455903ebc58a29da06b481f1e5c9775e -> 1bd1e604aebc0e1dd437cd294195332080596223 ([pr14810.101](https://github.com/hebasto/bitcoin/commits/pr14810.101) -> [pr14810.102](https://github.com/hebasto/bitcoin/commits/pr14810.102)) due to the conflict with #18402.

jonasschnelli commented:
I'm unsure if it is worth customising the label based on mouse events to achieve tabbing thought labels.
Also, on macOS, the selected color is different and almost not seeable.

master:
<img width="1046" alt="Bildschirmfoto 2020-05-29 um 09 09 20" src="https://user-images.githubusercontent.com/178464/83231777-59a99200-a18c-11ea-9600-e2e3601b3fa2.png">
<img width="905" alt="Bildschirmfoto 2020-05-29 um 09 09 27" src="https://user-images.githubusercontent.com/178464/83231789-5dd5af80-a18c-11ea-8af8-c3f89c5b437d.png">

this PR:
<img width="1046" alt="Bildschirmfoto 2020-05-29 um 09 10 44" src="https://user-images.githubusercontent.com/178464/83232022-bad16580-a18c-11ea-96d0-00a822fac41f.png">
<img width="905" alt="Bildschirmfoto 2020-05-29 um 09 11 07" src="https://user-images.githubusercontent.com/178464/83232024-bc029280-a18c-11ea-82fc-0f022fa490f8.png">

hebasto commented:
> This PR:
> 
>  - fixes a visual bug described in #14577

This fix is replaced by #19210.

>  - improves UX by enabling moving focus through fields using keyboard (**Tab** and **Shift+Tab** as usual) on Overview tab of the main window and on Info tab of the Debug window (other tabs leaved untouched as tab order for them seems a bit entangled). Inspired by **laanwj**'s [#14577 (comment)](https://github.com/bitcoin/bitcoin/pull/14577#issuecomment-438652059)

@jonasschnelli 

> I'm unsure if it is worth customising the label based on mouse events to achieve tabbing thought labels.
> Also, on macOS, the selected color is different and almost not seeable.

I have some ideas how to get the keyboard-only UX but not going to work on it right now.
So closing this PR.

Please label it "Up for grabs" :)
